File: app/src/main/java/com/ihadzhi/eatlimination/ui/ViewRecordingsAdapter.java
package com.ihadzhi.eatlimination.ui;

import android.content.Context;
import android.text.SpannableStringBuilder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.ihadzhi.eatlimination.R;
import com.ihadzhi.eatlimination.data.DietDao;
import com.ihadzhi.eatlimination.data.EatliminationDatabase;
import com.ihadzhi.eatlimination.data.FoodDao;
import com.ihadzhi.eatlimination.data.SymptomRecord;
import com.ihadzhi.eatlimination.databinding.RecordingsItemBinding;

import org.joda.time.DateTime;

import java.util.List;
import java.util.concurrent.Executor;

class ViewRecordingsAdapter extends RecyclerView.Adapter<ViewRecordingsAdapter.ViewRecordingsViewHolder> {

    @FunctionalInterface
    public interface RecordClickListener {
        void execute(SymptomRecord record);
    }

    private final Context context;
    private List<SymptomRecord> records;
    private RecordClickListener recordClickListener;

    public ViewRecordingsAdapter(Context context, RecordClickListener recordClickListener) {
        this.context = context;
        this.recordClickListener = recordClickListener;
    }

    @NonNull
    @Override
    public ViewRecordingsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        RecordingsItemBinding binding = RecordingsItemBinding.inflate(layoutInflater, parent, false);
        return new ViewRecordingsViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewRecordingsViewHolder holder, int position) {
        SymptomRecord record = records.get(position);
        holder.bind(record);
    }

    public void setRecords(List<SymptomRecord> records) {
        this.records = records;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return records != null ? records.size() : 0;
    }

    class ViewRecordingsViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        RecordingsItemBinding binding;
        private Executor executor;

        public ViewRecordingsViewHolder(RecordingsItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(SymptomRecord record) {
            binding.setRecord(record);
            itemView.setOnClickListener(this);
            DateTime dateTime = new DateTime(record.getTimestamp());
            String time = dateTime.getHourOfDay() + ":" + dateTime.getMinuteOfHour();
            StringBuilder description =
                    new StringBuilder(context.getString(R.string.symptom_description,
                            record.getValue(),
                            dateTime.toString("d MMM, yyyy"),
                            time));
            binding.symptomDescription.setText(description);
            if (record.getCategory() != null) {
                switch (record.getCategory()) {
                    case green:
                        binding.symptomCategory.setBackgroundResource(R.drawable.symptom_status_green_background);
                        break;
                    case yellow:
                        binding.symptomCategory.setBackgroundResource(R.drawable.symptom_status_yellow_background);
                        break;
                    case red:
                        binding.symptomCategory.setBackgroundResource(R.drawable.symptom_status_red_background);
                        break;
                }
            }
        }

        @Override
        public void onClick(View v) {
            if (recordClickListener != null && records != null && records.size() > 0) {
                recordClickListener.execute(records.get(getAdapterPosition()));
            }
        }

    }
}

File: app/src/main/java/com/ihadzhi/eatlimination/ui/ViewRecordingsAdapter.kt
package com.ihadzhi.eatlimination.ui

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.ihadzhi.eatlimination.R
import com.ihadzhi.eatlimination.data.SymptomRecord
import com.ihadzhi.eatlimination.data.SymptomRecord.SymptomCategory
import com.ihadzhi.eatlimination.databinding.RecordingsItemBinding
import com.ihadzhi.eatlimination.ui.ViewRecordingsAdapter.ViewRecordingsViewHolder
import org.joda.time.DateTime
import java.util.*
import java.util.concurrent.Executor

typealias RecordClickListener = (SymptomRecord) -> Unit

internal class ViewRecordingsAdapter(private val context: Context, private val recordClickListener: RecordClickListener?) : RecyclerView.Adapter<ViewRecordingsViewHolder>() {

    private var records: List<SymptomRecord>? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewRecordingsViewHolder {
        val layoutInflater = LayoutInflater.from(context)
        val binding = RecordingsItemBinding.inflate(layoutInflater, parent, false)
        return ViewRecordingsViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewRecordingsViewHolder, position: Int) {
        records?.get(position)?.let { record -> holder.bind(record) }
    }

    fun setRecords(records: List<SymptomRecord>?) {
        this.records = records
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return records?.size ?: 0
    }

    internal inner class ViewRecordingsViewHolder(var binding: RecordingsItemBinding) : RecyclerView.ViewHolder(binding.root), View.OnClickListener {
        private val executor: Executor? = null
        fun bind(record: SymptomRecord) {
            binding.record = record
            itemView.setOnClickListener(this)
            val calendar = Calendar.getInstance()
            calendar.time = record.timestamp
            val dateTime = DateTime(calendar[Calendar.YEAR], calendar[Calendar.MONTH] + 1,
                    calendar[Calendar.DAY_OF_MONTH], calendar[Calendar.HOUR_OF_DAY], calendar[Calendar.MINUTE], calendar[Calendar.SECOND])
            val description = StringBuilder(context.getString(R.string.symptom_description,
                    record.value,
                    dateTime.toString("d MMM, yyyy @ HH:MM")))
            binding.symptomDescription.text = description
            if (record.category != null) {
                when (record.category) {
                    SymptomCategory.green -> binding.symptomCategory.setBackgroundResource(R.drawable.symptom_status_green_background)
                    SymptomCategory.yellow -> binding.symptomCategory.setBackgroundResource(R.drawable.symptom_status_yellow_background)
                    SymptomCategory.red -> binding.symptomCategory.setBackgroundResource(R.drawable.symptom_status_red_background)
                }
            }
        }

        override fun onClick(v: View) {
            records?.get(adapterPosition)?.let { record -> recordClickListener?.invoke(record) }
        }
    }
}